File: easy/0501_find_mode_in_binary_search_tree/solution.cpp
#include "../../utils/cpp/nodes.hpp"
#include <stack>
#include <vector>

class Solution {
public:
  std::vector<int> findMode(utils::TreeNode *root) {
    std::vector<int> modes;
    int modeCount = 0;

    int currVal = -100001;
    int currCount = 0;

    utils::TreeNode *curr = root;
    std::stack<utils::TreeNode *> s;

    while (curr || !s.empty()) {
      while (curr) {
        s.push(curr);
        curr = curr->left;
      }

      utils::TreeNode *node = s.top();
      s.pop();

      if (currVal != node->val) {
        currVal = node->val;
        currCount = 1;
      } else
        ++currCount;

      if (currCount >= modeCount) {
        if (currCount > modeCount) {
          modes.clear();
          modeCount = currCount;
        }
        modes.push_back(node->val);
      }

      curr = node->right;
    }

    return modes;
  }
};

File: easy/0501_find_mode_in_binary_search_tree/solution.py
from collections import Counter
from typing import List, Optional

from utils.python.nodes import TreeNode


class Solution:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        vals_counter = Counter()

        stack = [root]
        while stack:
            node = stack.pop()

            vals_counter[node.val] += 1

            if node.left:
                stack.append(node.left)
            if node.right:
                stack.append(node.right)

        max_count = max(vals_counter.values())

        return [
            val for val, count in vals_counter.items() if count == max_count
        ]
